### PR TITLE
Reconcile Optimizations from S4

### LINKED ISSUE #16: Reconcile S4 Optimizer w/ Original Implementation
The original S4 repository indicates that optimization for S4 needs to be handled specially (https://github.com/HazyResearch/state-spaces/blob/feeab742e9c737c8e2b8b0e44d3efff4049f5847/example.py#L235).

Specifically:
- Fixed small learning rates for state space matrices, **with no weight decay** (we do not respect this with current AdamW).
- Larger learning rates & weight decay for other parameters.

### PR DESCRIPTION
**Pull request by siddk**
As part of quest to replicate CIFAR results (~89% from original repo [as reported here](https://github.com/HazyResearch/state-spaces/blob/feeab742e9c737c8e2b8b0e44d3efff4049f5847/example.py#L20)), implement some of the more nuanced details from the original codebase.

Includes:
- S4-specific optimization (low LR, no weight decay for S4 `B, C` params), regular LR + weight decay for others.
- Dropout2D (important) vs. 1D... not sure if this is totally right, but we want to drop out entire "channels" (timesteps)
- Fix default hyperparameters (learning rate, weight decay)
- LR Schedule (not exactly the same as original, but does matter)

Implementing the above found a few Jax/Flax oddities (especially in the relationship between Flax and Optax), as well as some other details in the original S4 code that seem important:
- In original S4 code, [the step size `dt` for discretization is learned](https://github.com/HazyResearch/state-spaces/blob/feeab742e9c737c8e2b8b0e44d3efff4049f5847/src/models/sequence/ss/kernel.py#L187). This alone shouldn't explain discrepancy in results. 
- By default Flax.Dropout w/ `deterministic = False` does [some weird rescaling stuff that shouldn't matter](https://flax.readthedocs.io/en/latest/_autosummary/flax.linen.Dropout.html#flax.linen.Dropout.deterministic) but haven't tested.
- Generally some [weird update logic I don't fully understand](https://github.com/HazyResearch/state-spaces/blob/feeab742e9c737c8e2b8b0e44d3efff4049f5847/src/models/sequence/ss/standalone/s4.py#L907).
- Parameter initialization is also different in original implementation.

General thoughts: I'm pretty confident that our implementation is educational and technically sound, and not missing anything major (Albert/Karan read through it at a high-level and said we hit the major points). 

But, if we want to replicate the SOTA results in the original repo, there's probably a million little things that we can tweak that will all add up to the results from the other repo... unclear how important they are in telling the story.

(Tentatively) Resolves #16.

### CODE REVIEW

**Comment by srush** [comment]
Thanks this is great. Just want to understand a couple other things

* What are our mnist results?
* Are we sure we don't have bugs in the main model? Have we replicated any smaller results?
* what's the current status with Albert/ Karen? They read the code?

File: README.md [hunk 1]
```diff
@@ -21,17 +21,47 @@ python -m s4.train --dataset mnist-classification --model s4 --epochs 10 --bsz 1
...
 # Default arguments (100 epochs for CIFAR)
 python -m s4.train --dataset cifar-classification --model s4 --epochs 100 --bsz 128 --d_model 128 --ssm_n 64
 
-# S4 replication (close enough to params in original repo's example.py -- no LR schedule though!)
+# S4 replication from central repository
 python -m s4.train --dataset cifar-classification --model s4 --epochs 100 --bsz 64 --d_model 512 --ssm_n 64
 ```
 
-(Default Arguments): Gets "best" 65.51% accuracy @ 46s/epoch on a TitanRTX
+Adding a Schedule:
```
Comment (srush) [inline README.md:49]: Not sure I understand these, is this comparing our implementation to the main one? 

Also is the speed relatively similar?

**Comment by srush** [comment]
Some more specific comments: 
* In original S4 code, the step size dt for discretization is learned. This alone shouldn't explain discrepancy in results. 

We can do this. Looks like they are updating it in log space. 

*  Generally some weird update logic I don't fully understand.

Yeah... looks like they are not training C, but are training A. Could see this making a difference.  Wonder if there is an easy way to do this in Flax (guessing not). 

* Parameter initialization is also different in original implementation.

This annoyingly might matter, I'll try to give a read.

**Comment by siddk** [comment]
Overview for this PR -- mostly attempt to get parity with original S4 results.

> What are our MNIST results?

We get 97.76% accuracy (in README) with a smaller model than in the [original codebase's example.py](https://github.com/HazyResearch/state-spaces/blob/feeab742e9c737c8e2b8b0e44d3efff4049f5847/example.py#L13). This is also *without* other bells/whistles -- no custom optimizer per layer, no dropout2D, no LR schedule, w/o fixed decay. 

If anything should be better now, but generally feels saturated.

> Are we sure we don't have bugs in the main model? Have we replicated any smaller results? What's the current status with Albert/ Karan?

The MNIST replication + the fact we can sample semi-decent completions with your existing code seem to indicate no "major" bugs with the model itself. Albert/Karan have both read through the code, no major issues with how we've laid things out.

They did mentions some things being important; Karan said CIFAR results were hanging in the 70s until they added Dropout2D, Albert by default always learned `log_dt` instead of having it fixed... but nothing in the main model that was glaringly wrong.

> Yeah... looks like they are not training C, but are training A. Could see this making a difference. Wonder if there is an easy way to do this in Flax (guessing not).

Yeah this + initialization would be where I suspect any meaningful discrepancies are. Not sure how quite to handle this... turning C training off is fine, we'd just drop the call to `self.param`... not sure about training A.

File: README.md (hunk 1 above)
Comment (siddk) [reply README.md:49]: Sorry this is unclear. Running with the same arguments as the main codebase (larger `d_model = 512`) takes forever, I found that running with a smaller model was nice to prototype changes a bit faster, and correlated pretty well. 

I just dropped the different results in the README -- "default args" = smaller model, "S4 replication" = bigger model (as apples-to-apples with original code as possible).

> Also is the speed relatively similar?

Albert/Karan report 7m20s per CIFAR epoch with their model on a T4 GPU, and ~2m30 on an A100. We're seeing 3m11s on a TitanRTX (in between the two GPUs)... Karan validated that speed checks out.
